src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ApplicationAssembliesProvider.cs

```diff
@@ -0,0 +1,256 @@
+                if (candidateEntry.Classification != DependencyClassification.Unknown)
+                {
+                    return candidateEntry.Classification;
+                }
+                else
+                {
+                    var classification = DependencyClassification.DoesNotReferenceMvc;
+                    foreach (var candidateDependency in candidateEntry.Library.Dependencies)
+                    {
+                        var dependencyClassification = ComputeClassification(candidateDependency.Name);
+                        if (dependencyClassification == DependencyClassification.ReferencesMvc ||
+                            dependencyClassification == DependencyClassification.MvcReference)
+                        {
+                            classification = DependencyClassification.ReferencesMvc;
+                            break;
+                        }
+                    }
+
+                    candidateEntry.Classification = classification;
+
+                    return classification;
+                }
+            }
+
+            public IEnumerable<RuntimeLibrary> GetCandidates()
+            {
+                foreach (var dependency in _runtimeDependencies)
+                {
+                    if (ComputeClassification(dependency.Key) == DependencyClassification.ReferencesMvc)
+                    {
+                        yield return dependency.Value.Library;
+                    }
+                }
+            }
+
+            private class Dependency
+            {
+                public Dependency(RuntimeLibrary library, DependencyClassification classification)
+                {
+                    Library = library;
+                    Classification = classification;
+                }
+
+                public RuntimeLibrary Library { get; }
+
+                public DependencyClassification Classification { get; set; }
+
+                public override string ToString()
+                {
+                    return $"Library: {Library.Name}, Classification: {Classification}";
+                }
+            }
+
+            private enum DependencyClassification
+            {
+                Unknown = 0,
+
+                /// <summary>
+                /// References (directly or transitively) one of the Mvc packages listed in
+                /// <see cref="ReferenceAssemblies"/>.
+                /// </summary>
+                ReferencesMvc = 1,
+
+                /// <summary>
+                /// Does not reference (directly or transitively) one of the Mvc packages listed by
+                /// <see cref="ReferenceAssemblies"/>.
+                /// </summary>
+                DoesNotReferenceMvc = 2,
+
+                /// <summary>
+                /// One of the references listed in <see cref="ReferenceAssemblies"/>.
+                /// </summary>
+                MvcReference = 3,
+            }
+        }
+
+        internal readonly struct AssemblyItem
```

Comment:
This smells like premature optimization to me

Reply:
The `struct`?